File: Week 2/vigenere.c
#include <cs50.h>
#include <stdio.h>
#include <ctype.h>
#include <string.h>

int checkIfNotValidKey(string key, int length);
int getIntForASCII(char letter);
void shiftAndPrintLetters(string text, int asciiArr[], int keyLength);

int main(int argc, string argv[])
{
    // Null value for key
    if (!argv[1]) {
        printf("Usage: ./vigenere keyword\n");
        return 1;
    }

    // Get length of second argument
    int len = strlen(argv[1]);

    // Not 2 arguments or not valid key
    if (argc !=2 || checkIfNotValidKey(argv[1], len))
    {
        printf("Usage: ./vigenere keyword\n");
        return 1;
    }

    // create array for ascii characters
    int asciiArray[len];

    // loop through the keyword and get ascii representation
    for (int i = 0; i<len; i++)
    {
        char letter = argv[1][i];
        int numb = getIntForASCII(letter);
        asciiArray[i] = numb;
    }

    //get plain text from user
    string plainText = get_string("plaintext: ");
    //send plain text, array, key length to new function
    shiftAndPrintLetters(plainText, asciiArray, len);
}

void shiftAndPrintLetters(string text, int asciiArr[], int keyLength)
{
    char defaultLowerChar = 'a';
    char defaultUpperChar = 'A';

    int lowerEndRange = 122;
    int upperEndRange = 90;

    int counter = 0;

    printf("ciphertext: ");
    for (int i=0, n = strlen(text); i<n; i++)
    {
        // Check if counter >= keyLength if so reset to 0.
        if (counter > keyLength - 1)
        {
          counter = 0;
        }

        // Not an alphabet, just print it.
        if (!isalpha(text[i]))
        {
            printf("%c", text[i]);
        }
        else if (isupper(text[i]))
        {
            int numb = text[i];
            if (numb + asciiArr[counter] <= upperEndRange)
            {
                printf("%c", text[i] + asciiArr[counter]);
            }
            else
            {
                int increase = (numb + asciiArr[counter]) - upperEndRange - 1;
                printf("%c", defaultUpperChar + (increase));
            }
            counter ++;
        }
        else
        {
            int numb = text[i];
            if (numb + asciiArr[counter] <= lowerEndRange)
            {
                printf("%c", text[i] + asciiArr[counter]);
            }
            else
            {
                int increase = (numb + asciiArr[counter]) - lowerEndRange - 1;
                printf("%c", defaultLowerChar + (increase));
            }
            counter ++;
        }
    }
   printf("\n");
}

int checkIfNotValidKey(string key, int length)
{
    for (int i = 0; i<length; i++)
    {
        if (!isalpha(key[i]))
        {
            return 1;
        }
    }
    return 0;
}

int getIntForASCII(char letter)
{
    string lowerCaseAlpha = "abcdefghijklmnopqrstuvwxyz";
    char lowerCase = tolower(letter);

    // Find corresponding index value
    for (int i = 0; i<26; i++)
    {
        if (lowerCaseAlpha[i] == lowerCase)
        {
            return i;
        }
    }
    return 1;
}

File: Week 6/vigenere.py
from sys import argv, exit
from cs50 import get_string


def main():
    if not argv[1] or len(argv) > 2:
        print("Usage: python vigenere.py k")
        exit(1)

    length = len(argv[1])

    # loop thru argv[1] and check each letter
    for i in range(length):
        if not argv[1][i].isalpha():
            print("Usage: python vigenere.py k")
            exit(1)

    letters = 'abcdefghijklmnopqrstuvwxyz'

    # 2 lists for upper and lower case letters
    lowerCaseList = []
    upperCaseList = []

    # list for index of ciper key
    cipherList = []
    # cipher counter, cannot exceed length of list
    counter = 0

    for i in range(26):
        lowerCaseList.append(letters[i])
        upperCaseList.append(letters[i].upper())

    # get cipherList as list of ints
    for i in range(length):
        for j in range(26):
            if argv[1][i].lower() == lowerCaseList[j]:
                cipherList.append(j)

    # prompt user
    plainText = get_string("plaintext: ")

    print("ciphertext: ", end="")

    # loop thru the plainText
    for i in range(len(plainText)):
        if counter > length - 1:
            counter = 0
        if not plainText[i].isalpha():
            print(plainText[i], end="")
        elif plainText[i].isupper():
            upper = shiftAndReturnLetter(cipherList[counter], plainText[i], upperCaseList)
            counter += 1
            print(upper, end="")
        else:
            lower = shiftAndReturnLetter(cipherList[counter], plainText[i], lowerCaseList)
            counter += 1
            print(lower, end="")

    print()


def shiftAndReturnLetter(key, letter, list):
    cipher = None
    for i in range(26):
        if letter == list[i]:
            cipher = (i + key) % 26
            break
    return list[cipher]


if __name__ == "__main__":
    main()
